drgn_tools/list_lru.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2024, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+
+"""
+LRU Lists
+------------
+
+Helper to work with LRU lists. LRU can be created to be memcg aware and
+ordered by NUMA node.
+
+The routines iterate through the specified LRU and on NUMA machines, the
+output keeps the entries ordered by NUMA node.
+
+The list_lru_from_memcg_node_for_each_entry() function allows the user to
+restrict the iteration by the memcg index when the list_lru is memcg_aware
+and the NUMA node identifier.
+
+list_lru_kmem_to_memcgidx() is a helper to find the mem_cgroup index
+from a list_lru kvm address. This helper will find the memcg of a list_lru
+address. This routine is only interested in slab allocated entries and does
+not check nor handle the MEMCG_DATA_KMEM case.
+"""
+
+import os
+from typing import Iterator, Union
+from drgn import cast, IntegerLike, NULL, Object, Program, Type
+from drgn.helpers.linux.list import list_empty, list_for_each_entry
+from drgn.helpers.linux.mm import compound_head, virt_to_page, page_to_virt
+from drgn.helpers.linux.nodemask import for_each_online_node, node_state
+from drgn.helpers.linux.xarray import xa_for_each, xa_load
+from drgn_tools.util import has_member
+
+
+__all__ = (
+    "list_lru_for_each_entry",
+    "list_lru_from_memcg_node_for_each_entry",
+    "list_lru_kmem_to_memcgidx",
+)
+
+
+def list_lru_for_each_entry(
+    prog: Program, type: Union[str, Type], lru: Object, member: str
+) -> Iterator[Object]:
+    """
+    Iterate over all of the entries in a list_lru.
+
+    :param prog: Kernel being debugged
+    :param type: Entry type.
+    :param lru: ``struct list_lru *``
+    :param member: Name of list node member in entry type.
+    :return: Iterator of ``type *`` objects.
+    """
+    if ((has_member(lru, "memcg_aware") and lru.memcg_aware) or
+        (has_member(lru.node, "memcg_lrus") and lru.node[0].memcg_lrus)) :
```

Comment:
sorry this has to be made smarter for uek7 memcg unaware lrus. 
I will  respin, test everywhere and make a new pull request